src/librustc/session/mod.rs

```diff
@@ -78,7 +78,7 @@ pub struct Session {
     /// Set of (LintId, Option<Span>, message) tuples tracking lint
     /// (sub)diagnostics that have been set once, but should not be set again,
     /// in order to avoid redundantly verbose output (Issue #24690).
-    pub one_time_diagnostics: RefCell<FxHashSet<(lint::LintId, Option<Span>, String)>>,
+    pub one_time_diagnostics: RefCell<FxHashSet<(Option<lint::LintId>, Option<Span>, String)>>,
```

Comment:
<rant>Normally, instead of making the first tuple field an option, I would suggest making the `FxHashSet` accept a new struct with three fields `diagnostic_id`, `span` and `message`, with `diagnostic_id` being a new enum that takes either a `LintId` or a `GateIssue`. This way we can keep expanding `one_time_diagnostics` without the risk of having multiple different diagnostics with the same span message and no lint_id, but being different errors, from being incorrectly stopped, but because of the `message` being there, what you're doing is correct and unlikely to be made incorrect with a subsequent change.</rant>